### PR TITLE
add mockstore and rawkv tests

### PR DESCRIPTION
**Pull request by disksing**
Signed-off-by: disksing <i@disksing.com>

### CODE REVIEW

**Comment by sunxiaoguang** [comment]
> Please take a look if you have spare time. @sunxiaoguang

Thanks, it's a good start for me to have better knowledge of it.

File: mockstore/mocktikv/mvcc_leveldb.go [hunk 1]
```diff
@@ -0,0 +1,1003 @@
...
+}
+
+type valueDecoder struct {
+	value     mvccValue
+	expectKey []byte
+}
+
+// valueDecoder decodes a mvcc value if iter key is expectKey.
+func (dec *valueDecoder) Decode(iter *Iterator) (bool, error) {
+	if iter.Error() != nil || !iter.Valid() {
```
Comment (rleungx) [inline mockstore/mocktikv/mvcc_leveldb.go]: It seems the validation has already been checked outside of this function except https://github.com/tikv/client-go/blob/7a23d6e7bf526a3b30847884c4005f0d3ddcc56f/mockstore/mocktikv/mvcc_leveldb.go#L545

File: mockstore/mocktikv/cluster.go [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+			return nil, context.Canceled
+		}
+		if s.meta.GetAddress() == addr {
+			return proto.Clone(s.meta).(*metapb.Store), nil
+		}
+	}
+	return nil, nil
+}
+
+// AddStore add a new Store to the cluster.
```
Comment (rleungx) [inline mockstore/mocktikv/cluster.go]: adds

File: mockstore/mocktikv/mvcc_leveldb.go (hunk 1 above)
Comment (disksing) [reply mockstore/mocktikv/mvcc_leveldb.go]: Copied from tidb directly. /cc @tiancaiamao Do you have any ideas?

Comment (tiancaiamao) [reply mockstore/mocktikv/mvcc_leveldb.go]: There is a `valid` field here  https://github.com/tikv/client-go/blob/7a23d6e7bf526a3b30847884c4005f0d3ddcc56f/mockstore/mocktikv/mvcc_leveldb.go#L125

If `iter.valid` is false, the `Decode` function should return as soon as possible. @rleungx 

The code is copied, I recommend not to modify it in this PR, if any bug is found, use a separate one to fix it.
b.t.w, tikv/client will be moved from tidb to this repo later? @disksing

Comment (disksing) [reply mockstore/mocktikv/mvcc_leveldb.go]: Yes. That's exactly I'm working on.

File: mockstore/mocktikv/cluster.go [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+
+func (c *Cluster) allocID() uint64 {
+	c.id++
+	return c.id
+}
+
+// GetAllRegions gets all the regions in the cluster.
+func (c *Cluster) GetAllRegions() []*Region {
+	regions := make([]*Region, 0, len(c.regions))
+	for _, region := range c.regions {
```
Comment (rleungx) [inline mockstore/mocktikv/cluster.go]: Do we need `RLock` here?

File: mockstore/mocktikv/cluster.go [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+	c.stores[storeID] = newStore(storeID, addr)
+}
+
+// GetRegion returns a Region's meta and leader ID.
+func (c *Cluster) GetRegion(regionID uint64) (*metapb.Region, uint64) {
+	c.RLock()
+	defer c.RUnlock()
+
+	r := c.regions[regionID]
+	if r == nil {
```
Comment (rleungx) [inline mockstore/mocktikv/cluster.go]: Maybe using `if r, ok := c.regions[regionID]; !ok` is better?

File: mockstore/mocktikv/cluster.go [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+func newPeerMeta(peerID, storeID uint64) *metapb.Peer {
+	return &metapb.Peer{
+		Id:      peerID,
+		StoreId: storeID,
+	}
+}
+
+func newRegion(regionID uint64, storeIDs, peerIDs []uint64, leaderPeerID uint64) *Region {
+	if len(storeIDs) != len(peerIDs) {
+		panic("len(storeIDs) != len(peerIds)")
```
Comment (rleungx) [inline mockstore/mocktikv/cluster.go]: `peerIDs`. BTW, is this a good panic message?

File: mockstore/mocktikv/cluster_manipulate.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (rleungx) [inline mockstore/mocktikv/cluster_manipulate.go]: should change all licenses to 2019?

**Comment by siddontang** [comment]
@disksing 

Maybe is it better to port and merge at first, then do the improvement later?

File: mockstore/mocktikv/cluster_manipulate.go (hunk 6 above)
Comment (disksing) [reply mockstore/mocktikv/cluster_manipulate.go]: No need to. The file is copied from tidb.

Comment (sunxiaoguang) [reply mockstore/mocktikv/cluster_manipulate.go]: Sadly, I have done the changes already...

Comment (disksing) [reply mockstore/mocktikv/cluster_manipulate.go]: I think it does not matter to change it or not to :)

File: mockstore/mocktikv/cluster.go [hunk 7]
```diff
@@ -0,0 +1,552 @@
...
+	return nil, nil
+}
+
+// Bootstrap creates the first Region. The Stores should be in the Cluster before
+// bootstrap.
+func (c *Cluster) Bootstrap(regionID uint64, storeIDs, peerIDs []uint64, leaderPeerID uint64) {
+	c.Lock()
+	defer c.Unlock()
+
+	if len(storeIDs) != len(peerIDs) {
```
Comment (rleungx) [inline mockstore/mocktikv/cluster.go]: Do we need to check it here? I find it has the same logic inside `newRegion`.

**Pull request merged**